docs/architecture/adr-060-baseapp++.md

```diff
@@ -0,0 +1,200 @@
+to store senders and their nonces.
+
+Transaction reaping will essentially happen via a two-phase approach:
+
+1. Reap one or more transactions from the priority queue and collect them into 
+   one of two buffers -- _valid_ and _invalid_.
+2. For transactions that DO NOT violate the nonce validation, they are included
+   in the _valid_ buffer.
+3. For transactions that DO violate the nonce validation, they are included in
+   the _invalid_ buffer.
+4. Continue this process until the desired number of valid transactions are
+   reaped or until the mempool is empty.
+5. Provide Tendermint the list of all transactions from the _valid_ buffer.
+6. Re-insert all transactions, from both buffers, back into mempool. This is to
+   ensure we do not discard transactions from the mempool in case `ProcessProposal`
+   fails.
+
+```go
+type PriorityMempool[T MempoolTx] struct {
+	queue   *PriorityQueue[MempoolTx]
+	senders map[string][]int64
+	// ...
+}
+```
+
+Previous discussions<sup>1</sup> have come to the agreement that Tendermint will
+perform a request to the application, via `PrepareProposalRequest`, with a certain
+amount of transactions reaped from Tendermint's local mempool. The exact amount
+of transactions reaped will be determined by a local operator configuration.
+This is referred to as the "one-shot approach" seen in discussions.
+
+When Tendermint reaps transactions from the local mempool and sends them to the
+application via `PrepareProposalRequest`, the application will have to evaluate
+the transactions. Specifically, it will need to inform Tendermint if it should
+reject and or include each transaction. Note, the application can even _replace_
+transactions entirely with other transactions.
+
+When evaluating transactions from `PrepareProposalRequest`, the application will
+ignore _all_ transactions sent to it in the request and instead reap up to
+`PrepareProposalRequest.max_tx_bytes` from it's own mempool. There is no need to
+execute the transactions for validity as they have already passed CheckTx.
+
+### `ProcessProposal`
+
+The `ProcessProposal` ABCI method is relatively straightforward. It is responsible
+for ensuring validity of the proposed block containing transactions that were
+selected from the `PrepareProposal` step. In order to check validity of the proposed
+block, we must iterate over the list of transactions and execute them using the
+same mode/execution strategy as `CheckTx`. In other words, we execute each transaction
+using the AnteHandler only -- no messages are executed. However, we cannot just
+execute `CheckTx` again, because `BaseApp` already has a modified `checkState` at
+this point. So when executing `ProcessProposal`, we create a similar branched
+state, `processProposalState`, off of `deliverState`. Using `processProposalState`
+we execute the AnteHandler for each transaction. Note, the `processProposalState`
+is never committed and is completely discarded after `ProcessProposal` completes.
```

Comment:
If the application always ignores transactions in the Tendermint mempool in favor of its own transactions, how does Tendermint know which transactions to evict from its own mempool? My understanding is that the only time a transaction is evicted from `clist_mempool` is if it fails check/recheck, or if it is committed in a block.

Reply:
Presumably [TxRecord.TxAction](https://github.com/tendermint/tendermint/blob/master/spec/abci%2B%2B/abci++_methods.md#txrecord) in `ResponsePrepareProposal.tx_records` informs Tendermint from the app.

Reply:
That makes sense in the case that `Tendermint` knows about the replacement transaction, but suppose the Application replaces all transactions with new transactions which did not originate from the Tendermint mempool.

Referring to this case:
> Note, the application can even _replace_ transactions entirely with other transactions.

Reply:
That makes sense. That would solve the case I am describing. It's unclear from this ADR whether a custom implementation of the Cosmos Mempool would provide the ability for the Application to decide whether transactions should be marked as `REMOVED` in the `tx_records` response to Tendermint.

Reply:
Agreed..

>If the Application considers that tx should not be proposed in this block, e.g., there are other transactions with higher priority, then it should not include it in tx_records. In this case, Tendermint will not remove tx from the mempool. The Application should be extra-careful, as abusing this feature may cause transactions to stay much longer than needed in the mempool.

Could the accumulation of many valid but low priority txs crash the app by filling the mempool?

Reply:
@alexanderbez Would it make sense for the `ReapMaxBytes` interface to have an additional return value of `[]TxAction` as well corresponding to the [TxAction](https://github.com/tendermint/tendermint/blob/master/spec/abci%2B%2B/abci++_methods.md#txaction) struct in the `tx_records` return value of [ResponsePrepareProposal](https://github.com/tendermint/tendermint/blob/master/spec/abci%2B%2B/abci++_methods.md#parameters-and-types-1). We could assume the slice to be ordered in the same manner as the `txRecords` argument, and under the hood Cosmos could combine these slices together to form the `tx_records` return value to Tendermint.

This would allow custom Cosmos mempool implementations to determine which transactions should be evicted from the mempool, which would be helpful for our usecase.

Reply:
Yeah, great point. I've updated `ReapMaxBytes`. Please lmk what you think :)